horizon/data_manager/update_operations.py

```diff
@@ -58,37 +75,77 @@ def _get_operations_for_update_role_assigment(
                             "role": role_key,
                             "resource": subject,
                         },
-                    )
+                    ),
                 )
 
 
 def _get_operations_for_update_user(
     user_key: str, data: JsonableValue
 ) -> Iterator[AnyOperation]:
     attributes = data.get("attributes", {})
-    yield InsertOperation(
-        fact=Fact(
-            type="users",
-            attributes={
-                "id": user_key,
-                "attributes": json.dumps(attributes),
-                # TODO remove the json.dumps after fixing the map[string]string issue in Go
-            },
+    if attributes:
+        yield InsertOperation(
+            fact=Fact(
+                type="users",
+                attributes={
+                    "id": user_key,
+                    "attributes": json.dumps(attributes),
+                    # TODO remove the json.dumps after fixing the map[string]string issue in Go
+                },
+            ),
+        )
+    else:
+        yield DeleteOperation(
+            fact=Fact(
+                type="users",
+                attributes={"id": user_key},
+            ),
+        )
+        yield DeleteOperation(
+            fact=Fact(
+                type="role_assignments",
+                attributes={"actor": user_key},
+            ),
         )
-    )
 
 
 def _get_operations_for_update_resource_instance(
     instance_key: str, data: JsonableValue
 ) -> Iterator[AnyOperation]:
     attributes = data.get("attributes", {})
-    yield InsertOperation(
-        fact=Fact(
-            type="resource_instance",
-            attributes={
-                "id": instance_key,
-                "attributes": json.dumps(attributes),
-                # TODO remove the json.dumps after fixing the map[string]string issue in Go
-            },
+    if attributes:
+        yield InsertOperation(
+            fact=Fact(
+                type="resource_instance",
+                attributes={
+                    "id": instance_key,
+                    "attributes": json.dumps(attributes),
+                    # TODO remove the json.dumps after fixing the map[string]string issue in Go
+                },
+            ),
+        )
+    else:
+        yield DeleteOperation(
+            fact=Fact(
+                type="resource_instance",
+                attributes={"id": instance_key},
+            ),
+        )
+        yield DeleteOperation(
+            fact=Fact(
+                type="relationship_tuples",
+                attributes={"object": instance_key},
+            ),
+        )
+        yield DeleteOperation(
+            fact=Fact(
+                type="relationship_tuples",
+                attributes={"subject": instance_key},
+            ),
+        )
+        yield DeleteOperation(
+            fact=Fact(
+                type="role_assignments",
+                attributes={"resource": instance_key},
+            ),
```

Comment:
why? a resource instance can exist without attributes? why remove it?

Reply:
When we "delete" an object from OPA we simply send it without any attributes, that's how the data update used to work. The attributes are not just the custom attributes, but contains the key as well, so if it doesn't exist then it's a delete operation

Reply:
ok, please add that in a comment.